shared/timeutils/timeutils.h

```diff
@@ -60,7 +60,7 @@ mp_uint_t timeutils_mktime_2000(mp_uint_t year, mp_int_t month, mp_int_t mday,
 
 static inline void timeutils_seconds_since_epoch_to_struct_time(uint64_t t, timeutils_struct_time_t *tm) {
     // TODO this will give incorrect results for dates before 2000/1/1
-    return timeutils_seconds_since_2000_to_struct_time(t - TIMEUTILS_SECONDS_1970_TO_2000, tm);
+    timeutils_seconds_since_2000_to_struct_time(t - TIMEUTILS_SECONDS_1970_TO_2000, tm);
```

Comment:
I was under the impression that this is valid C code, to return a void value in a void function.

Reply:
@dpgeorge C99 says not so:

> 6.8.6.4 The return statement
> Constraints
> A return statement with an expression shall not appear in a function whose return type is void. A return statement without an expression shall only appear in a function whose return type is void.

Reply:
Well, it does compile fine with `gcc -std=c99`.


> A return statement with an expression shall not appear in a function whose return type is void.

Technically this sentence doesn't apply here because `timeutils_seconds_since_2000_to_struct_time()` is not an expression, it doesn't return anything....?

Reply:
The reason I care is because at some point I wanted to replace all `mp_raise_XXX()` calls with `return mp_raise_XXX()` (to get rid of NLR), and that would sometimes introduce this case of return-void-expression.

Reply:
> Technically this sentence doesn't apply here because timeutils_seconds_since_2000_to_struct_time() is not an expression, it doesn't return anything....?

I might be misreading the standard, but I believe this is a *void expression (an expression that has type **void**)* as per section 6.3.2.2 void.

Section 6.5 Expressions is also illustrative, particularly the first point:

> 1. An expression is a sequence of operators and operands that specifies computation of a value, or that designates an object or a function, or that generates side effects, or that performs a combination thereof.

> Well, it does compile fine with gcc -std=c99.

Looks like it's enabled by `-pedantic`: https://gcc.godbolt.org/z/qzTejd66x

For my 2c, it doesn't seem like it'd necessarily be problematic to explicitly allow this pattern in MicroPython. Although maybe it's worth noting such exceptions in the coding style docs somewhere. Also, I don't know if there's a way to make MSVC allow it...

Reply:
> Looks like it's enabled by `-pedantic`

Yes you are right.

OK, I'm happy to change this then, so it doesn't return a void expression.

Side note: `-pedantic` raises a lot of warnings in the core code (eg when enabled for `bare-arm`).  I'm not sure we can eliminate all those warnings.

Reply:
> Side note: -pedantic raises a lot of warnings in the core code (eg when enabled for bare-arm). I'm not sure we can eliminate all those warnings.

I don't think we need to, those warnings are classed pedantic for a reason. :grin: